### PR TITLE
sherintg/vos on blob p2/DAOS_14363

### PR DESCRIPTION
**Pull request by sherintg**
- The phase-2 allocator will now support evictable and non-evictable memory buckets. The new allocator can be enabled using DAOS_MD_ON_SSD_MODE=3.
- Unit tests added to test the functionality.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/common/dav_v2/heap.c [hunk 1]
```diff
@@ -838,130 +1138,218 @@ heap_get_bestfit_block(struct palloc_heap *heap, struct bucket *b,
...
+		ERR("alloc failed");
+		return ENOMEM;
+	}
+
+	nzones    = heap_max_zone(store->stor_size);
+	start_off = ALIGN_UP(sizeof(struct heap_header), 4096);
+	for (i = 0; i < nzones;) {
+		nelems = ((nzones - i) > MAX_HEADER_FETCH) ? MAX_HEADER_FETCH : (nzones - i);
+		rc     = meta_fetch_batch(store, base, i * ZONE_MAX_SIZE + start_off,
+					  sizeof(struct zone), ZONE_MAX_SIZE, MAX_HEADER_FETCH);
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/dav_v2/heap.c]: I have fixed it now. allocation status, evictability and usage stats are now stored in a vec in zone0.

File: src/common/dav_v2/heap.c [hunk 2]
```diff
@@ -199,24 +496,25 @@ zone_calc_size_idx(uint32_t zone_id, unsigned max_zone, size_t heap_size)
...
 		.magic = ZONE_HEADER_MAGIC,
 	};
 
-	z->header = nhdr; /* write the entire header (8 bytes) at once */
+	z->header = nhdr; /* write the entire header at once */
+	if (is_evictable)
+		z->header.flags |= ZONE_EVICTABLE_MB;
 	mo_wal_persist(&heap->p_ops, &z->header, sizeof(z->header));
+
+	memblock_huge_init(heap, first_chunk_id, zone_id, size_idx - first_chunk_id);
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/dav_v2/heap.c:617]: mo_wal_persist only writes to the internal log. The wal commit of both zone header and chunk headers happen together.

File: src/common/mem.c [hunk 3]
```diff
@@ -2093,6 +2106,15 @@ cache_off2pg_off(struct umem_cache *cache, umem_off_t offset)
 	return (offset - cache->ca_base_off) & cache->ca_page_mask;
 }
 
+bool
+umem_cache_offisloaded(struct umem_store *store, umem_off_t offset)
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/mem.c:2091]: fixed it now

File: src/common/dav_v2/meta_io.c [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+/*
+ * meta_io.c -- IO to/from meta blob bypassing WAL.
+ */
+
+#include <errno.h>
+#include <daos/mem.h>
+
+/** Maximum number of sets of pages in-flight at a time */
+#define MAX_INFLIGHT_SETS 4
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/dav_v2/meta_io.c:12]: fetch_batch is not called now given that the evictability status is stored in zone0

File: src/common/dav_v2/meta_io.c [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+		sgl.sg_nr_out = iod.io_nr;
+
+		for (i = 0; i < iod.io_nr; i++) {
+			iod_region[i].sr_addr = start_off;
+			iod_region[i].sr_size = size;
+			start_off += hop_dist;
+		}
+		iod.io_regions = iod_region;
+
+		rc = store->stor_ops->so_write(store, &iod, &sgl);
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/dav_v2/meta_io.c:48]: meta blob empty flag is not really required for phase 2. It now loads only allocated non-evictable MBs and evictable MBs required for reboot as part of wal replay. Note that zone0 is always loaded

File: src/common/dav_v2/meta_io.c [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+	sgl.sg_nr             = 1;
+	sgl.sg_nr_out         = 1;
+	sgl.sg_iovs           = &sg_iov;
+	d_iov_set(&sg_iov, src, size);
+
+	D_ASSERT(store != NULL);
+	if (store->stor_ops->so_write == NULL)
+		return 0;
+
+	rc = store->stor_ops->so_write(store, &iod, &sgl);
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/dav_v2/meta_io.c:82]: Yes we can clear the meta blob empty flag, given that we only load populated MBs now during reboot.

File: src/common/dav_v2/tx.c [hunk 7]
```diff
@@ -1844,17 +1850,38 @@ dav_tx_publish_v2(struct dav_action *actv, size_t actvcnt)
...
-	return 0;
+	D_ASSERT((dav_tx_stage_v2() == DAV_TX_STAGE_NONE));
+
+	err = heap_get_evictable_mb(pop->do_heap, &mb_id);
+	if (err) {
+		errno = err;
+		return 0;
+	}
+
+	umem_cache_reserve(pop->do_store);
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/dav_v2/tx.c]: Fixed. Removed this

File: src/common/dav_v2/wal_tx.c [hunk 8]
```diff
@@ -427,6 +426,52 @@ struct umem_wal_tx_ops dav_wal_tx_ops = {
...
+	uint32_t                z_id = OFFSET_TO_ZID(off);
+	struct umem_cache_range rg   = {0};
+	int                     rc;
+
+	rg.cr_off  = GET_ZONE_OFFSET(z_id);
+	rg.cr_size = ((store->stor_size - rg.cr_off) > ZONE_MAX_SIZE)
+			 ? ZONE_MAX_SIZE
+			 : (store->stor_size - rg.cr_off);
+	rc         = umem_cache_load(store, &rg, 1, 0);
+	D_ASSERT(rc == 0);
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/dav_v2/wal_tx.c]: Fixed

File: src/common/mem.c [hunk 9]
```diff
@@ -1540,31 +1528,56 @@ bmem_atomic_free_v2(struct umem_instance *umm, umem_off_t umoff)
...
+	.mo_defer_free         = bmem_defer_free_v2,
+	.mo_cancel             = bmem_cancel_v2,
+	.mo_tx_publish         = bmem_tx_publish_v2,
+	.mo_atomic_copy        = bmem_atomic_copy_v2,
+	.mo_atomic_alloc       = bmem_atomic_alloc_v2,
+	.mo_atomic_free        = bmem_atomic_free_v2,
+	.mo_atomic_flush       = bmem_atomic_flush_v2,
+	.mo_allot_evictable_mb = bmem_allot_mb_evictable_v2,
+	.mo_get_mb_from_offset = bmem_get_mb_from_offset_v2,
+	.mo_get_mb_base_offset = bmem_get_mb_base_offset_v2,
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/mem.c]: Fixed. Replaced them with call to umem cache.

File: src/common/dav_v2/meta_io.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/dav_v2/meta_io.c:82]: I have modified the umem_cache code to honor phase-2 if mem_pages is set while calling umem_cache_alloc(). The phase-1 allocator will pase zero. The phase-2 allocator now honors mempages=metapages where its set max_ne_pages to mempages. Now its upto the vos to determine which allocator to use. Setting MD_ON_SSD_MODE=3 now will cause vos to setup phase-2 allocator right now.

File: src/common/mem.c [hunk 10]
```diff
@@ -2213,7 +2249,7 @@ umem_cache_free(struct umem_store *store)
 static inline unsigned int
 cache_mode(struct umem_cache *cache)
 {
-	return (cache->ca_md_pages == cache->ca_mem_pages) ? 1 : 2;
+	return (cache->ca_max_ne_pages == cache->ca_mem_pages) ? 1 : 2;
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/mem.c]: Fixed the logic now. I have modified the umem_cache code to honor phase-2 if mem_pages is set while calling umem_cache_alloc(). The phase-1 allocator will pase zero. The phase-2 allocator now honors mempages=metapages where its set max_ne_pages to mempages. Now its upto the vos to determine which allocator to use. Setting MD_ON_SSD_MODE=3 now will cause vos to setup phase-2 allocator right now.

File: src/common/mem.c [hunk 11]
```diff
@@ -3143,13 +3181,15 @@ cache_evict_page(struct umem_cache *cache, bool for_sys)
...
 {
-	uint32_t	page_nr;
+	uint32_t	page_nr = 0;
 
-	/* Few free pages are always reserved for potential non-evictable zone grow */
-	D_ASSERT(cache->ca_pgs_stats[UMEM_PG_STATS_NONEVICTABLE] <= cache->ca_max_ne_pages);
-	page_nr = cache->ca_max_ne_pages - cache->ca_pgs_stats[UMEM_PG_STATS_NONEVICTABLE];
-	if (page_nr > UMEM_CACHE_RSRVD_PAGES)
-		page_nr = UMEM_CACHE_RSRVD_PAGES;
+	if (!cache->ca_early_boot) {
```
Comment (NiuYawei) [inline src/common/mem.c:3186]: If we rename the 'ca_early_boot' to 'ca_replay_done' and set the 'ca_replay_done' flag in umem_cache_post_replay(), the new API of umem_cache_set_early_boot() could be removed.

File: src/common/mem.c [hunk 12]
```diff
@@ -2213,7 +2239,7 @@ umem_cache_free(struct umem_store *store)
 static inline unsigned int
 cache_mode(struct umem_cache *cache)
 {
-	return (cache->ca_md_pages == cache->ca_mem_pages) ? 1 : 2;
+	return cache->ca_mode;
```
Comment (NiuYawei) [inline src/common/mem.c:2242]: I don't think this change is necessary, no matter if DAV v1 or v2 is used, the cache mode is always decided by "md_pages == mem_pages", right?

File: src/common/dav_v2/dav_iface.c [hunk 13]
```diff
@@ -18,63 +18,84 @@
...
 
-/*
- * get_uuid_lo -- (internal) evaluates XOR sum of least significant
- * 8 bytes with most significant 8 bytes.
- */
-static inline uint64_t
-get_uuid_lo(uuid_t uuid)
-{
-	uint64_t uuid_lo = 0;
+D_CASSERT(UMEM_CACHE_PAGE_SZ == ZONE_MAX_SIZE);
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/dav_v2/dav_iface.c]: Removed reference to UMEM_CACHE_PAGE_SZ

File: src/common/dav_v2/dav_iface.c [hunk 14]
```diff
@@ -83,118 +104,139 @@ dav_obj_open_internal(int fd, int flags, size_t sz, const char *path, struct ume
...
 	hdl->do_store = store;
+	hdl->p_ops.umem_store = store;
+
 	if (hdl->do_store->stor_priv == NULL) {
 		D_ERROR("meta context not defined. WAL commit disabled for %s\n", path);
 	} else {
-		num_pages = (sz + UMEM_CACHE_PAGE_SZ - 1) >> UMEM_CACHE_PAGE_SZ_SHIFT;
-		rc = umem_cache_alloc(store, UMEM_CACHE_PAGE_SZ, num_pages, 0, 0, 0, base,
-				      NULL, NULL);
+		rc = umem_cache_alloc(store, UMEM_CACHE_PAGE_SZ, hzl.nzones_heap, hzl.nzones_cache,
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/dav_v2/dav_iface.c]: Fixed

File: src/common/dav_v2/heap_layout.h [hunk 15]
```diff
@@ -10,20 +10,21 @@
...
 #define HEAP_MINOR 0
 
-#define MAX_CHUNK (UINT16_MAX - 7) /* has to be multiple of 8 */
+#define MAX_CHUNK             63
 #define CHUNK_BASE_ALIGNMENT 1024
-#define CHUNKSIZE ((size_t)1024 * 256)	/* 256 kilobytes */
+#define CHUNKSIZE             ((size_t)1024 * 260) /* 260 kilobytes */
 #define MAX_MEMORY_BLOCK_SIZE (MAX_CHUNK * CHUNKSIZE)
 #define HEAP_SIGNATURE_LEN 16
 #define HEAP_SIGNATURE "MEMORY_HEAP_HDR\0"
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/dav_v2/heap_layout.h:23]: In V1, the heap starts at offset 4k while on v2 it starts from offset 0.

File: src/common/dav_v2/heap.c [hunk 16]
```diff
@@ -838,159 +1276,206 @@ heap_get_bestfit_block(struct palloc_heap *heap, struct bucket *b,
...
+		D_CRIT("heap: invalid header's checksum\n");
+		return -1;
+	}
 
-	for (i = 0; i < MAX_ALLOCATION_CLASSES; ++i) {
-		if (default_zset->buckets[i] == NULL)
-			continue;
-		bucket_locked_delete(default_zset->buckets[i]);
+	if ((hdr->major != HEAP_MAJOR) || (hdr->minor != HEAP_MINOR)) {
+		D_ERROR("Version mismatch of heap layout\n");
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/dav_v2/heap.c:1377]: Fixed the the check for minor to greater than or equal to.

File: src/common/dav_v2/heap.c [hunk 17]
```diff
@@ -1041,106 +1529,224 @@ heap_boot(struct palloc_heap *heap, void *heap_start, uint64_t heap_size,
...
 {
+	int      nzones;
+	uint64_t heap_size = store->stor_size;
+
 	if (heap_size < HEAP_MIN_SIZE)
 		return EINVAL;
 
-	VALGRIND_DO_MAKE_MEM_UNDEFINED(heap_start, heap_size);
+	if (store->stor_priv == NULL)
+		return 0;
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/dav_v2/heap.c]: Removed this check and added asserts at appropriate location. This was added in phase 1 for early testing. Now its not required.

File: src/common/mem.c [hunk 18]
```diff
@@ -3375,6 +3415,36 @@ cache_rgs2pgs(struct umem_cache *cache, struct umem_cache_range *ranges, int ran
...
+void
+umem_cache_set_early_boot(struct umem_store *store, bool mode)
+{
+	struct umem_cache *cache = store->cache;
+
+	cache->ca_early_boot = mode;
+}
+
+void
+umem_cache_update_nonevictable_stats(struct umem_store *store)
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/mem.c]: Fixed

File: src/common/mem.c [hunk 19]
```diff
@@ -3375,6 +3415,36 @@ cache_rgs2pgs(struct umem_cache *cache, struct umem_cache_range *ranges, int ran
...
+{
+	struct umem_cache     *cache = store->cache;
+	int                    cnt   = 0;
+	int                    idx;
+	struct umem_page_info *pinfo;
+
+	pinfo = (struct umem_page_info *)&cache->ca_pages[cache->ca_md_pages];
+	for (idx = 0; idx < cache->ca_mem_pages; idx++) {
+		if (pinfo[idx].pi_mapped == 0)
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/mem.c]: Fixed

File: src/common/dav_v2/tx.c [hunk 20]
```diff
@@ -456,6 +456,7 @@ lw_tx_begin(dav_obj_t *pop)
 	int			 rc;
 	uint64_t		 wal_id;
 
+	umem_cache_reserve(pop->do_store);
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/dav_v2/tx.c]: Fixed

File: src/common/dav_v2/wal_tx.c [hunk 21]
```diff
@@ -437,20 +492,24 @@ dav_wal_replay_cb(uint64_t tx_id, struct umem_action *act, void *arg)
...
 			"ACT_COPY txid=%lu, (p,o)=%lu,%lu size=%lu\n",
 			tx_id,
 			act->ac_copy.addr / PAGESIZE, act->ac_copy.addr % PAGESIZE,
 			act->ac_copy.size);
-		off = act->ac_copy.addr;
-		dst = base + off;
+		off  = act->ac_copy.addr;
 		src = (void *)&act->ac_copy.payload;
 		size = act->ac_copy.size;
+		dst  = dav_wal_replay_heap_off2ptr(dav_hdl, off);
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/dav_v2/wal_tx.c:473]: Removed the new logic of pinning the page during replay and retained the old behavior. Fixed cache evict code to reset the tx in flight and tx associated with last checkpoint for the page when a page is evicted.

File: src/common/dav_v2/wal_tx.c [hunk 22]
```diff
@@ -427,6 +426,62 @@ struct umem_wal_tx_ops dav_wal_tx_ops = {
...
+{
+	struct dav_wal_replay_cache *dwrc  = (struct dav_wal_replay_cache *)dav_hdl->do_cb_wa;
+	struct umem_store           *store = dav_hdl->do_store;
+	int                          i;
+
+	if (tx_id == dwrc->last_txid)
+		return;
+
+	if (dwrc->last_txid)
+		umem_cache_commit(store, dwrc->last_txid);
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/dav_v2/wal_tx.c]: Removed the new logic of pinning the page during replay and retained the old behavior. Fixed cache evict code to reset the tx in flight and tx associated with last checkpoint for the page when a page is evicted.

File: src/common/dav_v2/wal_tx.c [hunk 23]
```diff
@@ -427,6 +426,62 @@ struct umem_wal_tx_ops dav_wal_tx_ops = {
 	.wtx_act_next = wal_tx_act_next,
 };
 
+struct dav_wal_replay_cache {
+	uint64_t                last_txid;
+	int                     capacity;
+	int                     cur_pos;
+	struct umem_pin_handle *pinhdl[];
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/dav_v2/wal_tx.c]: Removed the new logic of pinning the page during replay and retained the old behavior. Fixed cache evict code to reset the tx in flight and tx associated with last checkpoint for the page when a page is evicted.

File: src/common/mem.c [hunk 24]
```diff
@@ -2120,6 +2132,20 @@ umem_cache_ptr2off(struct umem_store *store, void *ptr)
...
+bool
+umem_cache_ptrisvalid(struct umem_store *store, void *ptr)
+{
+	struct umem_cache *cache = store->cache;
+	uint32_t           idx;
+
+	if (ptr < cache->ca_base)
+		return false;
+	idx = (ptr - cache->ca_base) >> cache->ca_page_shift;
+	if (idx >= cache->ca_mem_pages)
```
[the remark this answers is not in the document]
Comment (sherintg) [reply src/common/mem.c]: Removed this function as part of rework.

File: src/common/mem.c (hunk 11 above)
Comment (sherintg) [reply src/common/mem.c]: Fixed

File: src/common/mem.c (hunk 12 above)
Comment (sherintg) [reply src/common/mem.c:2219]: As discussed offline, the mode is decided by the allocator type.

**Review by NiuYawei — APPROVED** [review]

**Review by wangshilong — APPROVED** [review]

**Pull request merged**